src/libcore/alloc.rs

```diff
@@ -579,6 +579,17 @@ pub unsafe trait GlobalAlloc {
     }
 }
 
+/// A hack so the default impl can condition on the associated type. This `Err`
+/// type ought to just live in the `Alloc` trait.
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub trait AllocHelper {
+
+    /// The type of any errors thrown by the allocator, customarily
+    /// either `AllocErr`, for when error recovery is allowed, or `!`
+    /// to signify that all errors will result in .
+    type Err = AllocErr;
```

Comment:
Also, why does it have a default error type?

Reply:
Optional backwards compat with existing instances.

Reply:
We don't need backwards compatibility.